File: src/js/fable-core/List.ts
import List from "./ListClass";
import { ofArray } from "./ListClass";
import { groupBy as mapGroupBy } from "./Map";
import { map as seqMap } from "./Seq";
import { fold as seqFold } from "./Seq";
import { foldBack as seqFoldBack } from "./Seq";
import { toList as seqToList } from "./Seq";

export default List;
export { ofArray };

export function append<T>(xs: List<T>, ys: List<T>) {
  return seqFold((acc, x) => new List<T>(x, acc), ys, reverse(xs));
}

export function choose<T, U>(f: (x: T) => U, xs: List<T>) {
  const r = seqFold((acc, x) => {
    const y = f(x);
    return y != null ? new List<U>(y, acc) : acc;
  }, new List<U>(), xs);
  return reverse(r);
}

export function collect<T, U>(f: (x: T) => List<U>, xs: List<T>) {
  return seqFold((acc, x) => append(acc, f(x)), new List<U>(), xs);
}
// TODO: should be xs: Iterable<List<T>>

export function concat<T>(xs: List<List<T>>) {
  return collect((x) => x, xs);
}

export function filter<T>(f: (x: T) => boolean, xs: List<T>) {
  return reverse(seqFold((acc, x) => f(x) ? new List<T>(x, acc) : acc, new List<T>(), xs));
}

export function where<T>(f: (x: T) => boolean, xs: List<T>) {
  return filter(f, xs);
}

export function initialize<T>(n: number, f: (i: number) => T) {
  if (n < 0) {
    throw new Error("List length must be non-negative");
  }
  let xs = new List<T>();
  for (let i = 1; i <= n; i++) {
    xs = new List<T>(f(n - i), xs);
  }
  return xs;
}

export function map<T, U>(f: (x: T) => U, xs: List<T>) {
  return reverse(seqFold((acc: List<U>, x: T) => new List<U>(f(x), acc), new List<U>(), xs));
}

export function mapIndexed<T, U>(f: (i: number, x: T) => U, xs: List<T>) {
  return reverse(seqFold((acc, x, i) => new List<U>(f(i, x), acc), new List<U>(), xs));
}

export function indexed<T>(xs: List<T>) {
  return mapIndexed((i, x) => [i, x] as [number, T], xs);
}

export function partition<T>(f: (x: T) => boolean, xs: List<T>) {
  return seqFold((acc, x) => {
    const lacc = acc[0];
    const racc = acc[1];
    return f(x) ? [new List<T>(x, lacc), racc] : [lacc, new List<T>(x, racc)];
  }, [new List<T>(), new List<T>()], reverse(xs));
}

export function replicate<T>(n: number, x: T) {
  return initialize(n, () => x);
}

export function reverse<T>(xs: List<T>) {
  return seqFold((acc, x) => new List<T>(x, acc), new List<T>(), xs);
}

export function singleton<T>(x: T) {
  return new List<T>(x, new List<T>());
}

export function slice<T>(lower: number, upper: number, xs: List<T>) {
  const noLower = (lower == null);
  const noUpper = (upper == null);
  return reverse(seqFold((acc, x, i) =>
    (noLower || lower <= i) && (noUpper || i <= upper) ? new List<T>(x, acc) : acc, new List<T>(), xs));
}
/* ToDo: instance unzip() */

export function unzip<T1, T2>(xs: List<[T1, T2]>) {
  return seqFoldBack((xy, acc) =>
    [new List<T1>(xy[0], acc[0]), new List<T2>(xy[1], acc[1])] as [List<T1>, List<T2>],
    xs,
    [new List<T1>(), new List<T2>()] as [List<T1>, List<T2>]);
}
/* ToDo: instance unzip3() */

export function unzip3<T1, T2, T3>(xs: List<[T1, T2, T3]>) {
  return seqFoldBack((xyz, acc) =>
    [new List<T1>(xyz[0], acc[0]), new List<T2>(xyz[1], acc[1]), new List<T3>(xyz[2], acc[2]),
    ] as [List<T1>, List<T2>, List<T3>],
    xs,
    [new List<T1>(), new List<T2>(), new List<T3>(),
    ] as [List<T1>, List<T2>, List<T3>]);
}

export function groupBy<T, K>(f: (x: T) => K, xs: List<T>): List<[K, List<T>]> {
  return seqToList(seqMap((k) => [k[0], seqToList(k[1])], mapGroupBy(f, xs))) as List<[K, List<T>]>;
}

export function splitAt<T>(index: number, xs: List<T>): [List<T>, List<T>] {
  if (index < 0) {
    throw new Error("The input must be non-negative.");
  }
  let i = 0;
  let last = xs;
  const first = new Array(index);
  while (i < index) {
    if (last.tail == null) {
      throw new Error("The input sequence has an insufficient number of elements.");
    }
    first[i] = last.head;
    last = last.tail;
    i++;
  }
  return [ofArray(first), last];
}

File: src/js/fable-core/List.fs
module List

[<AutoOpen>]
module private JS =
    open Fable.Core.JsInterop

    // TODO!!!: Pass the appropriate zero/one for the type from Fable
    let inline nativeZero<'T> : 'T = !!0
    let inline nativeOne<'T> : 'T = !!1

let rec foldIndexedAux f i acc = function
   | [] -> acc
   | x::xs -> foldIndexedAux f (i+1) (f i acc x) xs

let foldIndexed<'a,'acc> f (seed:'acc) (xs: 'a list) =
   foldIndexedAux f 0 seed xs

let fold<'a,'acc> f (seed:'acc) (xs: 'a list) =
   foldIndexed (fun _ acc x -> f acc x) seed xs

let reverse xs =
   fold (fun acc x -> x::acc) [] xs

let foldBack<'a,'acc> f (xs: 'a list) (seed:'acc) =
   fold (fun acc x -> f x acc) seed (reverse xs)

let rec foldIndexed2Aux f i acc bs cs =
   match bs, cs with
   | [], [] -> acc
   | x::xs, y::ys -> foldIndexed2Aux f (i+1) (f i acc x y) xs ys
   | _ -> invalidOp "Lists had different lengths"

let foldIndexed2<'a, 'b, 'acc> f (seed:'acc) (xs: 'a list) (ys: 'b list) =
   foldIndexed2Aux f 0 seed xs ys

let fold2<'a, 'b, 'acc> f (seed:'acc) (xs: 'a list) (ys: 'b list) =
   foldIndexed2 (fun _ acc x y -> f acc x y) seed xs ys

let foldBack2<'a, 'b, 'acc> f (xs: 'a list) (ys: 'b list) (seed:'acc) =
   fold2 f seed (reverse xs) (reverse ys)

let rec foldIndexed3Aux f i acc bs cs ds =
   match bs, cs, ds with
   | [], [], [] -> acc
   | x::xs, y::ys, z::zs -> foldIndexed3Aux f (i+1) (f i acc x y z) xs ys zs
   | _ -> invalidOp "Lists had different lengths"

let foldIndexed3<'a, 'b, 'c, 'acc> f (seed:'acc) (xs: 'a list) (ys: 'b list) (zs: 'c list) =
   foldIndexed3Aux f 0 seed xs ys zs

let fold3<'a, 'b, 'c, 'acc> f (seed:'acc) (xs: 'a list) (ys: 'b list) (zs: 'c list) =
   foldIndexed3 (fun _ acc x y z -> f acc x y z) seed xs ys zs

let scan<'a, 'acc> f (seed:'acc) (xs: 'a list) =
   fold (fun acc x ->
      match acc with
      | [] -> failwith "never"
      | y::_ -> f y x::acc) [seed] xs
   |> reverse

let scanBack<'a, 'acc> f (xs: 'a list) (seed:'acc) =
   scan (fun acc x -> f x acc) seed (reverse xs)
   |> reverse

let length xs =
   fold (fun acc _ -> acc + 1) 0 xs

let append xs ys =
   fold (fun acc x -> x::acc) ys (reverse xs)

let collect f xs =
   fold (fun acc x -> append (f x) acc) [] (reverse xs)

let map f xs =
   fold (fun acc x -> (f x::acc)) [] xs
   |> reverse

let mapIndexed f xs =
   foldIndexed (fun i acc x -> f i x::acc) [] xs
   |> reverse

let map2 f xs ys =
   fold2 (fun acc x y -> f x y::acc) [] xs ys
   |> reverse

let mapIndexed2 f xs ys =
   foldIndexed2 (fun i acc x y  -> f i x y:: acc) [] xs ys
   |> reverse

let map3 f xs ys zs =
   fold3 (fun acc x y z -> f x y z::acc) [] xs ys zs
   |> reverse

let mapIndexed3 f xs ys zs =
   foldIndexed3 (fun i acc x y z -> f i x y z:: acc) [] xs ys zs
   |> reverse

let iterate f xs =
   fold (fun () x -> f x) () xs

let iterate2 f xs ys =
   fold2 (fun () x y -> f x y) () xs ys

let iterateIndexed f xs =
   foldIndexed (fun i () x -> f i x) () xs

let iterateIndexed2 f xs ys =
   foldIndexed2 (fun i () x y -> f i x y) () xs ys

let ofArray xs =
   Array.foldBack (fun x acc -> x::acc) xs []

let toArray xs =
   let size = length xs
   let ys = Array.zeroCreate size
   iterateIndexed (fun i x -> ys.[i] <- x) xs
   ys

let empty<'a> : 'a list = []

let isEmpty = function
   | [] -> true
   | _ -> false

let rec tryPickIndexedAux f i = function
   | [] -> None
   | x::xs ->
      let result = f i x
      match result with
      | Some _ -> result
      | None -> tryPickIndexedAux f (i+1) xs

let tryPickIndexed f xs =
   tryPickIndexedAux f 0 xs

let tryPick f xs =
   tryPickIndexed (fun _ x -> f x) xs

let pick f xs =
   match tryPick f xs with
   | None -> invalidOp "List did not contain any matching elements"
   | Some x -> x

let tryFindIndexed f xs =
   tryPickIndexed (fun i x -> if f i x then Some x else None) xs

let tryFind f xs =
   tryPickIndexed (fun _ x -> if f x then Some x else None) xs

let findIndexed f xs =
   match tryFindIndexed f xs with
   | None -> invalidOp "List did not contain any matching elements"
   | Some x -> x

let find f xs =
   findIndexed (fun _ x -> f x) xs

let tryFindIndex f xs =
   tryPickIndexed (fun i x -> if f x then Some i else None) xs

let findIndex f xs =
   match tryFindIndex f xs with
   | None -> invalidOp "List did not contain any matching elements"
   | Some x -> x

let item n xs =
   findIndexed (fun i _ -> n = i) xs

let filter f xs =
   foldBack (fun x acc ->
      if f x then x::acc
      else acc) xs []

let partition f xs =
   fold (fun (lacc, racc) x ->
      if f x then x::lacc, racc
      else lacc,x::racc) ([],[]) xs

let choose f xs =
   fold (fun acc x ->
      match f x with
      | Some y -> y:: acc
      | None -> acc) [] xs |> reverse

let initialize n f =
   let mutable xs = []
   for i = 1 to n do xs <- f (n - i):: xs
   xs

let replicate n x =
   initialize n (fun _ -> x)

let reduce f = function
   | [] -> invalidOp "List was empty"
   | h::t -> fold f h t

let reduceBack f = function
   | [] -> invalidOp "List was empty"
   | h::t -> foldBack f t h

let forAll f xs =
   fold (fun acc x -> acc && f x) true xs

let forAll2 f xs ys =
   fold2 (fun acc x y -> acc && f x y) true xs ys

let rec exists f = function
   | [] -> false
   | x::xs -> f x || exists f xs

let rec exists2 f bs cs =
   match bs, cs with
   | [], [] -> false
   | x::xs, y::ys -> f x y || exists2 f xs ys
   | _ -> invalidOp "Lists had different lengths"

let unzip xs =
   foldBack (fun (x, y) (lacc, racc) -> x::lacc, y::racc) xs ([],[])

let unzip3 xs =
   foldBack (fun (x, y, z) (lacc, macc, racc) -> x::lacc, y::macc, z::racc) xs ([],[],[])

let zip xs ys =
   map2 (fun x y -> x, y) xs ys

let zip3 xs ys zs =
   map3 (fun x y z -> x, y, z) xs ys zs

let sort xs =
   xs
   |> toArray
   |> Array.sort
   |> ofArray

let sortWith f xs =
   xs
   |> toArray
   |> Array.sortWith f
   |> ofArray

let sum (xs: ^a list) : ^a =
   fold (fun acc x -> acc + x) nativeZero xs

let sumBy (f:^a -> ^b) (xs: ^a list) : ^b =
   fold (fun acc x -> acc + f x) nativeZero xs

let maxBy f xs =
   reduce (fun x y -> if f y > f x then y else x) xs

let max xs =
   reduce max xs

let minBy f xs =
   reduce (fun x y -> if f y > f x then x else y) xs

let min xs =
   reduce min xs

// TODO: Pass add and divide function for non-number types
let average (zs: ^a list) : ^a =
   let total = sum zs
   let count = sumBy (fun _ -> nativeOne< ^a >) zs
   total / count

let averageBy (g: ^a -> ^b ) (zs: ^a list) : ^b =
   let total = sumBy g zs
   let count = sumBy (fun _ -> nativeOne< ^a >) zs
   total / count

let permute f xs =
   xs
   |> toArray
   |> Array.permute f
   |> ofArray

let sortBy f xs =
   let ys = xs |> toArray
   Array.sortInPlaceBy f ys
   ys |> ofArray

// TODO: Is there a more efficient algorithm?
let take i xs =
  let rec takeInner i acc xs =
      match i, xs with
      | 0, _ -> reverse acc
      | _, [] -> failwith "The input sequence has an insufficient number of elements."
      | _, x::xs -> takeInner (i - 1) (x::acc) xs
  match i, xs with
  | i, _ when i < 0 -> failwith "The input must be non-negative."
  | 0, _ -> []
  | 1, x::_ -> [x]
  | i, xs -> takeInner i [] xs

let skip i xs =
  let rec skipInner i xs =
      match i, xs with
      | 0, _ -> xs
      | _, [] -> failwith "The input sequence has an insufficient number of elements."
      | _, _::xs -> skipInner (i - 1) xs
  match i, xs with
  | i, _ when i < 0 -> failwith "The input must be non-negative."
  | 0, _ -> xs
  | 1, _::xs -> xs
  | i, xs -> skipInner i xs

let toSeq (xs: _ list) =
    seq {
        let mutable xs = xs
        while not xs.IsEmpty do
            yield xs.Head
            xs <- xs.Tail
    }